### PR TITLE
[SYCL][OpenMP] Correctly handle __bf16 intrinsics using deferred diagnostics

### PR DESCRIPTION
**Pull request by elizabethandrews**
Commits 1 and 2 are cherry-picks from llvm.org - https://reviews.llvm.org/D141375 and https://reviews.llvm.org/rGc10615e4a94fc8dec65a48a6eb8f7efccc3fb1fc.

Commit 3 reverts header changes made in intel/llvm.

### CODE REVIEW

**Review by smanna12 — APPROVED** [review]

**Comment by bader** [comment]
Please, follow https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#pull-request recommendations for PR decoration. E.g. PR title "Correctly handle __bf16 intrinsics using deferred diagnostics" -> "[SYCL][OpenMP] Use deferred diagnostics to handle __bf16".

**Review by Fznamznon — APPROVED** [review]

**Comment by elizabethandrews** [comment]
@intel/llvm-gatekeepers this patch is ready for review and merge. Failing test is unrelated to this PR

**Comment by steffenlarsen** [comment]
Failed Tests (1):
  SYCL :: dword_atomic_smoke.cpp
Since fixed in https://github.com/intel/llvm-test-suite/pull/1552

File: clang/test/SemaSYCL/bf16.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+void host_ok(void) {
+  __bf16 A;
+}
+
+int main()
+{  host_ok();
+  __bf16 var; // expected-note {{'var' defined here}}
+  kernel<class variables>([=]() {
+    (void)var; // expected-error {{'var' requires 16 bit size '__bf16' type support, but target 'spir64' does not support it}}
+    int B = sizeof(__bf16);
```
Comment (premanandrao) [inline clang/test/SemaSYCL/bf16.cpp:17]: (Sorry, this is late.  My notifications weren't working.)

Maybe a nit.  Why is this declaration of B needed?  Or is this showing that it does not get a diagnostic?

Comment (elizabethandrews) [reply clang/test/SemaSYCL/bf16.cpp:17]: Yes I added it to show we don't get a diagnostic when used in unevaluated context

**Pull request merged**